src/contracts/middleware/BLSRegistryCoordinatorWithIndices.sol

```diff
@@ -0,0 +1,555 @@
+        
+        // deregister the operator from the BLSPubkeyRegistry
+        blsPubkeyRegistry.deregisterOperator(operator, quorumNumbers, pubkey);
+
+        // deregister the operator from the StakeRegistry
+        stakeRegistry.deregisterOperator(operatorId, quorumNumbers);
+
+        // deregister the operator from the IndexRegistry
+        indexRegistry.deregisterOperator(operatorId, quorumNumbers, operatorIdsToSwap);
+
+        // set the toBlockNumber of the operator's quorum bitmap update
+        _operatorIdToQuorumBitmapHistory[operatorId][operatorQuorumBitmapHistoryLengthMinusOne].nextUpdateBlockNumber = uint32(block.number);
+        
+        // if it is not a complete deregistration, add a new quorum bitmap update
+        if (!completeDeregistration) {
+            _operatorIdToQuorumBitmapHistory[operatorId].push(QuorumBitmapUpdate({
+                updateBlockNumber: uint32(block.number),
+                nextUpdateBlockNumber: 0,
+                quorumBitmap: quorumBitmapBeforeUpdate & ~uint192(quorumsToRemoveBitmap) // this removes the quorumsToRemoveBitmap from the quorumBitmapBeforeUpdate
+            }));
+        } else {
+            // @notice Registrant must continue to serve until the latest block at which an active task expires. this info is used in challenges
+            uint32 latestServeUntilBlock = serviceManager.latestServeUntilBlock();
+
+            // record a stake update unbonding the operator after `latestServeUntilBlock`
+            // serviceManager.recordLastStakeUpdateAndRevokeSlashingAbility(operator, latestServeUntilBlock);
+            // set the status of the operator to DEREGISTERED
+            _operators[operator].status = OperatorStatus.DEREGISTERED;
+
+            emit OperatorDeregistered(operator, operatorId);
+        }
+    }
+
+    /// @notice verifies churnApprover's signature on operator churn approval and increments the churnApprover nonce
+    function _verifyChurnApproverSignatureOnOperatorChurnApproval(bytes32 registeringOperatorId, OperatorKickParam[] memory operatorKickParams, SignatureWithSaltAndExpiry memory signatureWithSaltAndExpiry) internal {
+        // make sure the salt hasn't been used already
+        require(!isChurnApproverSaltUsed[signatureWithSaltAndExpiry.salt], "BLSRegistryCoordinatorWithIndices._verifyChurnApproverSignatureOnOperatorChurnApproval: churnApprover salt already used");
+        require(signatureWithSaltAndExpiry.expiry >= block.timestamp, "BLSRegistryCoordinatorWithIndices._verifyChurnApproverSignatureOnOperatorChurnApproval: churnApprover signature expired");        
+        EIP1271SignatureUtils.checkSignature_EIP1271(churnApprover, calculateOperatorChurnApprovalDigestHash(registeringOperatorId, operatorKickParams, signatureWithSaltAndExpiry.salt, signatureWithSaltAndExpiry.expiry), signatureWithSaltAndExpiry.signature);
+        // set salt used to true
+        isChurnApproverSaltUsed[signatureWithSaltAndExpiry.salt] = true;
```

Comment:
this is some CIE design, rather than CEI.
although reentrancy by a trusted role is less important to protect against, we should have the order of operations flipped here (i.e. mark salt as spent, then potentially call to external contract)